Build is broken

Seems the build is broken; depending upon a SNAPSHOT dependency:

```
[ERROR] Failed to execute goal on project openshift-jenkins-sync-plugin: Could not resolve dependencies for project io.fabric8:openshift-jenkins-sync-plugin:hpi:0.1-SNAPSHOT: Could not find artifact io.fabric8:kubernetes-model:jar:1.0.44-SNAPSHOT in repo.jenkins-ci.org (http://repo.jenkins-ci.org/public/) -> [Help 1]
```

When setting this locally to `io.fabric8:kubernetes-model:jar:1.0.44` I see a compilation error:

```
ERROR] /home/alr/business/jboss/git/fabric8io/openshift-jenkins-sync-plugin/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java:[29,38] cannot find symbol
[ERROR] symbol:   class JenkinsPipelineBuildStrategy
[ERROR] location: package io.fabric8.openshift.api.model
```

Dependency of my working on #56

--- a/pom.xml
+++ b/pom.xml
@@ -43,8 +43,7 @@
          ~ hpi-plugin.version: The HPI Maven Plugin version used by the plugin..
          ~ stapler-plugin.version: The Stapler Maven plugin version required by the plugin.
     -->
-    <openshift-client.version>1.3.92</openshift-client.version>
-    <kubernetes-model.version>1.0.44-SNAPSHOT</kubernetes-model.version>
+    <openshift-client.version>1.3.96</openshift-client.version>
   </properties>
 
   <name>OpenShift Jenkins Sync Plugin</name>
@@ -122,12 +121,6 @@
       <groupId>io.fabric8</groupId>
       <artifactId>openshift-client</artifactId>
       <version>${openshift-client.version}</version>
-    </dependency>
-
-    <dependency>
-      <groupId>io.fabric8</groupId>
-      <artifactId>kubernetes-model</artifactId>
-      <version>${kubernetes-model.version}</version>
     </dependency>
 
     <dependency>

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigWatcher.java
@@ -18,12 +18,15 @@ package io.fabric8.jenkins.openshiftsync;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import hudson.model.Job;
 import hudson.security.ACL;
+import io.fabric8.kubernetes.api.model.Status;
 import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
 import io.fabric8.kubernetes.client.Watcher;
 import io.fabric8.openshift.api.model.BuildConfig;
 import io.fabric8.openshift.api.model.BuildConfigList;
 import jenkins.model.Jenkins;
 import jenkins.security.NotReallyRoleSensitiveCallable;
+import jenkins.util.Timer;
 import org.apache.tools.ant.filters.StringInputStream;
 import org.jvnet.hudson.reactor.ReactorException;
 
@@ -32,13 +35,16 @@ import javax.xml.transform.stream.StreamSource;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static io.fabric8.jenkins.openshiftsync.BuildConfigToJobMapper.mapBuildConfigToJob;
 import static io.fabric8.jenkins.openshiftsync.JenkinsUtils.xstream2;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getOpenShiftClient;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.isJenkinsBuildConfig;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.parseResourceVersion;
+import static java.net.HttpURLConnection.HTTP_GONE;
 
 /**
  * Watches {@link BuildConfig} objects in OpenShift and for WorkflowJobs we ensure there is a
@@ -46,16 +52,80 @@ import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.parseResourceVersi
  */
 public class BuildConfigWatcher implements Watcher<BuildConfig> {
   private final Logger logger = Logger.getLogger(getClass().getName());
-  private final String defaultNamespace;
+  private final String namespace;
+  private Watch buildConfigWatch;
 
   public BuildConfigWatcher(String defaultNamespace) {
-    this.defaultNamespace = defaultNamespace;
+    this.namespace = defaultNamespace;
+  }
+
+  public void start() {
+    final BuildConfigList buildConfigs;
+    if (namespace != null && !namespace.isEmpty()) {
+      buildConfigs = getOpenShiftClient().buildConfigs().inNamespace(namespace).list();
+      buildConfigWatch = getOpenShiftClient().buildConfigs().inNamespace(namespace).withResourceVersion(buildConfigs.getMetadata().getResourceVersion()).watch(this);
+    } else {
+      buildConfigs = getOpenShiftClient().buildConfigs().inAnyNamespace().list();
+      buildConfigWatch = getOpenShiftClient().buildConfigs().inAnyNamespace().withResourceVersion(buildConfigs.getMetadata().getResourceVersion()).watch(this);
+    }
+
+    // lets process the initial state
+    logger.info("Now handling startup build configs!!");
+    // lets do this in a background thread to avoid errors like:
+    //  Tried proxying io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration to support a circular dependency, but it is not an interface.
+    Runnable task = new Runnable() {
+      @Override
+      public void run() {
+        logger.info("Waiting for Jenkins to be started");
+        while (true) {
+          Jenkins jenkins = Jenkins.getInstance();
+          if (jenkins != null) {
+            if (jenkins.isAcceptingTasks()) {
+              break;
+            }
+          }
+          try {
+            Thread.sleep(500);
+          } catch (InterruptedException e) {
+            // ignore
+          }
+        }
+        logger.info("loading initial BuildConfigs resources");
+
+        try {
+          onInitialBuildConfigs(buildConfigs);
+          logger.info("loaded initial BuildConfigs resources");
+        } catch (Exception e) {
+          logger.log(Level.SEVERE, "Failed to load initial BuildConfigs: " + e, e);
+        }
+      }
+    };
+    // lets give jenkins a while to get started ;)
+    Timer.get().schedule(task, 500, TimeUnit.MILLISECONDS);
+  }
+
+  public void stop() {
+    if (buildConfigWatch != null) {
+      buildConfigWatch.close();
+    }
   }
 
   @Override
   public void onClose(KubernetesClientException e) {
     if (e != null) {
       logger.warning(e.toString());
+
+      if (e.getStatus() != null && e.getStatus().getCode() == HTTP_GONE) {
+        stop();
+        start();
+      }
+    }
+  }
+
+  @Override
+  public void errorReceived(Status status) {
+    if (status != null) {
+      logger.warning("Watch error received: " + status.toString());
     }
   }
 
@@ -99,8 +169,8 @@ public class BuildConfigWatcher implements Watcher<BuildConfig> {
       ACL.impersonate(ACL.SYSTEM, new NotReallyRoleSensitiveCallable<Void, IOException>() {
         @Override
         public Void call() throws IOException {
-          String jobName = OpenShiftUtils.jenkinsJobName(buildConfig, defaultNamespace);
-          Job jobFromBuildConfig = mapBuildConfigToJob(buildConfig, defaultNamespace);
+          String jobName = OpenShiftUtils.jenkinsJobName(buildConfig, namespace);
+          Job jobFromBuildConfig = mapBuildConfigToJob(buildConfig, namespace);
           if (jobFromBuildConfig == null) {
             return null;
           }

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java
@@ -24,8 +24,6 @@ import hudson.model.Run;
 import hudson.model.TaskListener;
 import hudson.model.listeners.RunListener;
 import io.fabric8.openshift.api.model.Build;
-import io.fabric8.openshift.api.model.BuildStatus;
-import io.fabric8.openshift.api.model.BuildStatusBuilder;
 import jenkins.util.Timer;
 import org.jenkinsci.plugins.workflow.job.WorkflowRun;
 import org.joda.time.DateTime;
@@ -35,7 +33,6 @@ import org.kohsuke.stapler.DataBoundConstructor;
 
 import javax.annotation.Nonnull;
 import java.io.IOException;
-import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CopyOnWriteArraySet;
 import java.util.concurrent.TimeUnit;
@@ -204,57 +201,53 @@ public class BuildSyncRunListener extends RunListener<Run> {
     }
 
     final Build build = cause.getBuild();
-    Map<String, String> annotations = build.getMetadata().getAnnotations();
-    annotations.put(ANNOTATION_JENKINS_STATUS_JSON, json);
 
     String rootUrl = OpenShiftUtils.getJenkinsURL(getOpenShiftClient(), defaultNamespace);
     String logsUrl = joinPaths(rootUrl, run.getUrl(), "/consoleText");
 
-    if (!annotations.containsKey(ANNOTATION_JENKINS_BUILD_URI)) {
-      annotations.put(ANNOTATION_JENKINS_BUILD_URI, run.getUrl());
-    }
-    if (!annotations.containsKey(ANNOTATION_JENKINS_LOG_URL)) {
-      annotations.put(ANNOTATION_JENKINS_LOG_URL, logsUrl);
-    }
     String name = build.getMetadata().getName();
 
-    build.setStatus(updateBuildStatus(build.getStatus(), run));
+    String phase = runToBuildPhase(run);
+
+    long started = getStartTime(run);
+    String startTime = null;
+    String completionTime = null;
+    if (started > 0) {
+      DateTime dateTime = new DateTime(startTime);
+      startTime = dateFormatter.print(dateTime);
+
+      long duration = getDuration(run);
+      if (duration > 0) {
+        dateTime = new DateTime(started + duration);
+        completionTime = dateFormatter.print(dateTime);
+      }
+    }
 
     if (logger.isLoggable(Level.FINE)) {
       logger.fine("generated build in namespace " + defaultNamespace + " with name: " + name + " phase: " + build.getStatus().getPhase() + " data: " + build);
     }
 
     logger.info("replacing build in namespace " + defaultNamespace + " with name: " + name + " phase: " + build.getStatus().getPhase());
-    getOpenShiftClient().builds().inNamespace(defaultNamespace).withName(name).replace(build);
+    getOpenShiftClient().builds().inNamespace(defaultNamespace).withName(name).edit()
+      .editMetadata()
+        .addToAnnotations(ANNOTATION_JENKINS_STATUS_JSON, json)
+        .addToAnnotations(ANNOTATION_JENKINS_BUILD_URI, run.getUrl())
+        .addToAnnotations(ANNOTATION_JENKINS_LOG_URL, logsUrl)
+      .endMetadata()
+      .editStatus()
+        .withPhase(phase)
+        .withStartTimestamp(startTime)
+        .withCompletionTimestamp(completionTime)
+      .endStatus()
+    .done();
   }
 
-  private BuildStatus updateBuildStatus(BuildStatus buildStatus, Run run) {
-    BuildStatusBuilder builder;
-    if (buildStatus != null) {
-      builder = new BuildStatusBuilder(buildStatus);
-      if (buildStatus.getPhase() != null) {
-        if (buildStatus.getPhase().equals(BuildPhases.NEW)
-          || buildStatus.getPhase().equals(BuildPhases.PENDING)
-          || buildStatus.getPhase().equals(BuildPhases.RUNNING)) {
-          builder.withPhase(runToBuildPhase(run));
-        }
-      }
-    } else {
-      builder = new BuildStatusBuilder();
-    }
+  private long getStartTime(Run run) {
+    return run.getStartTimeInMillis();
+  }
 
-    long startTime = run.getStartTimeInMillis();
-    if (startTime > 0) {
-      DateTime dateTime = new DateTime(startTime);
-      builder.withStartTimestamp(dateFormatter.print(dateTime));
-    }
-    long duration = run.getDuration();
-    if (duration > 0) {
-      DateTime dateTime = new DateTime(startTime + duration);
-      builder.withCompletionTimestamp(dateFormatter.print(dateTime));
-    }
-
-    return builder.build();
+  private long getDuration(Run run) {
+    return run.getDuration();
   }
 
   private String runToBuildPhase(Run run) {

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java
@@ -17,35 +17,103 @@ package io.fabric8.jenkins.openshiftsync;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import hudson.model.Job;
+import io.fabric8.kubernetes.api.model.Status;
 import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
 import io.fabric8.kubernetes.client.Watcher;
 import io.fabric8.openshift.api.model.Build;
 import io.fabric8.openshift.api.model.BuildConfig;
 import io.fabric8.openshift.api.model.BuildList;
-import io.fabric8.openshift.client.OpenShiftClient;
+import jenkins.model.Jenkins;
+import jenkins.util.Timer;
 import org.apache.commons.lang.StringUtils;
 
 import java.io.IOException;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static io.fabric8.jenkins.openshiftsync.BuildPhases.NEW;
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.cancelOpenShiftBuild;
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getOpenShiftClient;
+import static java.net.HttpURLConnection.HTTP_GONE;
 
 public class BuildWatcher implements Watcher<Build> {
   private static final Logger logger = Logger.getLogger(BuildWatcher.class.getName());
+  private final String namespace;
+  private Watch newBuildsWatch;
+  private Watch runningBuildsWatch;
 
-  private final OpenShiftClient openShiftClient;
+  public BuildWatcher(String defaultNamespace) {
+    this.namespace = defaultNamespace;
+  }
 
-  public BuildWatcher(OpenShiftClient openShiftClient) {
-    this.openShiftClient = openShiftClient;
+  public void start() {
+    final BuildList builds;
+    if (namespace != null && !namespace.isEmpty()) {
+      builds = getOpenShiftClient().builds().inNamespace(namespace).withField("status", BuildPhases.NEW).list();
+      newBuildsWatch = getOpenShiftClient().builds().inNamespace(namespace).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(this);
+      runningBuildsWatch = getOpenShiftClient().builds().inNamespace(namespace).withField("status", BuildPhases.RUNNING).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(this);
+    } else {
+      builds = getOpenShiftClient().builds().inAnyNamespace().withField("status", BuildPhases.NEW).list();
+      newBuildsWatch = getOpenShiftClient().builds().inAnyNamespace().withResourceVersion(builds.getMetadata().getResourceVersion()).watch(this);
+      runningBuildsWatch = getOpenShiftClient().builds().inAnyNamespace().withField("status", BuildPhases.RUNNING).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(this);
+    }
+
+    // lets process the initial state
+    logger.info("Now handling startup builds!!");
+    // lets do this in a background thread to avoid errors like:
+    //  Tried proxying io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration to support a circular dependency, but it is not an interface.
+    Runnable task = new Runnable() {
+      @Override
+      public void run() {
+        logger.info("Waiting for Jenkins to be started");
+        while (true) {
+          Jenkins jenkins = Jenkins.getInstance();
+          if (jenkins != null) {
+            if (jenkins.isAcceptingTasks()) {
+              break;
+            }
+          }
+          try {
+            Thread.sleep(500);
+          } catch (InterruptedException e) {
+            // ignore
+          }
+        }
+        logger.info("loading initial Builds resources");
+
+        try {
+          onInitialBuilds(builds);
+          logger.info("loaded initial Builds resources");
+        } catch (Exception e) {
+          logger.log(Level.SEVERE, "Failed to load initial Builds: " + e, e);
+        }
+      }
+    };
+    // lets give jenkins a while to get started ;)
+    Timer.get().schedule(task, 500, TimeUnit.MILLISECONDS);
+  }
+
+  public void stop() {
+    if (newBuildsWatch != null) {
+      newBuildsWatch.close();
+    }
+    if (runningBuildsWatch != null) {
+      runningBuildsWatch.close();
+    }
   }
 
   @Override
   public void onClose(KubernetesClientException e) {
     if (e != null) {
       logger.warning(e.toString());
+
+      if (e.getStatus() != null && e.getStatus().getCode() == HTTP_GONE) {
+        stop();
+        start();
+      }
     }
   }
 
@@ -82,6 +150,13 @@ public class BuildWatcher implements Watcher<Build> {
     }
   }
 
+  @Override
+  public void errorReceived(Status status) {
+    if (status != null) {
+      logger.warning("Watch error received: " + status.toString());
+    }
+  }
+
   private void buildModified(Build build) {
     if (Boolean.TRUE.equals(build.getStatus().getCancelled())) {
       Job job = getJobFromBuild(build);
@@ -109,7 +184,7 @@ public class BuildWatcher implements Watcher<Build> {
     if (StringUtils.isEmpty(buildConfigName)) {
       return null;
     }
-    BuildConfig buildConfig = openShiftClient.buildConfigs().inNamespace(build.getMetadata().getNamespace()).withName(buildConfigName).get();
+    BuildConfig buildConfig = getOpenShiftClient().buildConfigs().inNamespace(build.getMetadata().getNamespace()).withName(buildConfigName).get();
     if (buildConfig == null) {
       return null;
     }

--- a/src/main/java/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration.java
+++ b/src/main/java/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration.java
@@ -16,18 +16,11 @@
 package io.fabric8.jenkins.openshiftsync;
 
 import hudson.Extension;
-import io.fabric8.kubernetes.client.Watch;
-import io.fabric8.openshift.api.model.BuildConfigList;
-import io.fabric8.openshift.api.model.BuildList;
 import jenkins.model.GlobalConfiguration;
-import jenkins.model.Jenkins;
-import jenkins.util.Timer;
 import net.sf.json.JSONObject;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.StaplerRequest;
 
-import java.util.concurrent.TimeUnit;
-import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getNamespaceOrUseDefault;
@@ -44,11 +37,9 @@ public class GlobalPluginConfiguration extends GlobalConfiguration {
 
   private String namespace;
 
-  private transient Watch buildConfigWatch;
+  private transient BuildWatcher buildWatcher;
 
-  private transient Watch buildWatch;
-
-  private transient Watch runningBuildsWatch;
+  private transient BuildConfigWatcher buildConfigWatcher;
 
   @DataBoundConstructor
   public GlobalPluginConfiguration(boolean enable, String server, String namespace) {
@@ -111,14 +102,11 @@ public class GlobalPluginConfiguration extends GlobalConfiguration {
     logger.info("using default kubernetes namespace: " + namespace);
 
     if (!enabled) {
-      if (buildConfigWatch != null) {
-        buildConfigWatch.close();
+      if (buildConfigWatcher != null) {
+        buildConfigWatcher.stop();
       }
-      if (buildWatch != null) {
-        buildWatch.close();
-      }
-      if (runningBuildsWatch != null) {
-        runningBuildsWatch.close();
+      if (buildWatcher != null) {
+        buildWatcher.stop();
       }
       OpenShiftUtils.shutdownOpenShiftClient();
       return;
@@ -127,63 +115,12 @@ public class GlobalPluginConfiguration extends GlobalConfiguration {
       OpenShiftUtils.initializeOpenShiftClient(server);
       this.namespace = getNamespaceOrUseDefault(namespace, getOpenShiftClient());
 
-      final BuildConfigWatcher buildConfigWatcher = new BuildConfigWatcher(namespace);
-      final BuildWatcher buildWatcher = new BuildWatcher(getOpenShiftClient());
-      final BuildConfigList buildConfigs;
-      final BuildList builds;
-      if (namespace != null && !namespace.isEmpty()) {
-        buildConfigs = getOpenShiftClient().buildConfigs().inNamespace(namespace).list();
-        buildConfigWatch = getOpenShiftClient().buildConfigs().inNamespace(namespace).withResourceVersion(buildConfigs.getMetadata().getResourceVersion()).watch(buildConfigWatcher);
-        builds = getOpenShiftClient().builds().inNamespace(namespace).withField("status", BuildPhases.NEW).list();
-        buildWatch = getOpenShiftClient().builds().inNamespace(namespace).withField("status", BuildPhases.NEW).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(buildWatcher);
-        runningBuildsWatch = getOpenShiftClient().builds().inNamespace(namespace).withField("status", BuildPhases.RUNNING).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(buildWatcher);
-      } else {
-        buildConfigs = getOpenShiftClient().buildConfigs().inAnyNamespace().list();
-        buildConfigWatch = getOpenShiftClient().buildConfigs().inAnyNamespace().withResourceVersion(buildConfigs.getMetadata().getResourceVersion()).watch(buildConfigWatcher);
-        builds = getOpenShiftClient().builds().inAnyNamespace().withField("status", BuildPhases.NEW).list();
-        buildWatch = getOpenShiftClient().builds().inAnyNamespace().withField("status", BuildPhases.NEW).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(buildWatcher);
-        runningBuildsWatch = getOpenShiftClient().builds().inAnyNamespace().withField("status", BuildPhases.RUNNING).withResourceVersion(builds.getMetadata().getResourceVersion()).watch(buildWatcher);
-      }
+      buildWatcher = new BuildWatcher(namespace);
+      buildWatcher.start();
+      buildConfigWatcher = new BuildConfigWatcher(namespace);
+      buildConfigWatcher.start();
 
-      // lets process the initial state
-      logger.info("Now handling startup build configs!!");
-      // lets do this in a background thread to avoid errors like:
-      //  Tried proxying io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration to support a circular dependency, but it is not an interface.
-      Runnable task = new Runnable() {
-        @Override
-        public void run() {
-          logger.info("Waiting for Jenkins to be started");
-          while (true) {
-            Jenkins jenkins = Jenkins.getInstance();
-            if (jenkins != null) {
-              if (jenkins.isAcceptingTasks()) {
-                break;
-              }
-            }
-            try {
-              Thread.sleep(500);
-            } catch (InterruptedException e) {
-              // ignore
-            }
-          }
-          logger.info("loading initial BuildConfigs resources");
 
-          try {
-            buildConfigWatcher.onInitialBuildConfigs(buildConfigs);
-            logger.info("loaded initial BuildConfigs resources");
-          } catch (Exception e) {
-            logger.log(Level.SEVERE, "Failed to load initial BuildConfigs: " + e, e);
-          }
-          try {
-            buildWatcher.onInitialBuilds(builds);
-            logger.info("loaded initial Builds resources");
-          } catch (Exception e) {
-            logger.log(Level.SEVERE, "Failed to load initial Builds: " + e, e);
-          }
-        }
-      };
-      // lets give jenkins a while to get started ;)
-      Timer.get().schedule(task, 500, TimeUnit.MILLISECONDS);
     }
   }
 

